File: 3640-maximum-frequency-of-an-element-after-performing-operations-ii/3640-maximum-frequency-of-an-element-after-performing-operations-ii.py
from collections import Counter, defaultdict
import bisect

class Solution:
    def maxFrequency(self, nums: list[int], k: int, numOperations: int) -> int:
        freq = Counter(nums)
        events = defaultdict(int)
        for v in nums:
            L, R = v - k, v + k
            events[L] += 1
            events[R + 1] -= 1

        items = sorted(events.items())
        pos, pref = [], []
        cur = 0
        for p, d in items:
            pos.append(p)
            cur += d
            pref.append(cur)

        max_cover = max(pref)
        ans = min(max_cover, numOperations)

        for v, cntEq in freq.items():
            i = bisect.bisect_right(pos, v) - 1
            cover = pref[i] if i >= 0 else 0
            cand = min(cover, cntEq + numOperations)
            ans = max(ans, cand)
        return ans

File: 3640-maximum-frequency-of-an-element-after-performing-operations-ii/3640-maximum-frequency-of-an-element-after-performing-operations-ii.java
import java.util.*;

class Solution {
    public int maxFrequency(int[] nums, long k, long numOperations) {
        Map<Long,Integer> freq = new HashMap<>();
        for (int x: nums) freq.put((long)x, freq.getOrDefault((long)x,0)+1);

        TreeMap<Long, Long> events = new TreeMap<>();
        for (int x: nums) {
            long v = x;
            long L = v - k, R = v + k;
            events.put(L, events.getOrDefault(L, 0L) + 1L);
            events.put(R + 1, events.getOrDefault(R + 1, 0L) - 1L);
        }

        List<Long> pos = new ArrayList<>();
        List<Long> pref = new ArrayList<>();
        long cur = 0;
        for (Map.Entry<Long,Long> e: events.entrySet()) {
            pos.add(e.getKey());
            cur += e.getValue();
            pref.add(cur);
        }

        long maxCover = 0;
        for (long x: pref) maxCover = Math.max(maxCover, x);

        long ans = Math.min(maxCover, numOperations);

        for (Map.Entry<Long,Integer> e: freq.entrySet()) {
            long v = e.getKey();
            int cntEq = e.getValue();
            int idx = Collections.binarySearch(pos, v);
            long cover;
            if (idx >= 0) cover = pref.get(idx);
            else {
                int ins = -idx - 1;
                cover = (ins == 0) ? 0 : pref.get(ins - 1);
            }
            long cand = Math.min(cover, cntEq + numOperations);
            ans = Math.max(ans, cand);
        }

        return (int)ans;
    }
}
